racker/salus-telemetry-monitor-management
# Add exclude resources option during monitor binding

# Resolves

Part of https://jira.rax.io/browse/SALUS-695

# What

Adds an `excludedResourceIds` field to monitors which allows for a monitor to be selectively excluded from binding to the given resource that would be normally be selected by labels.

# How

Tweaks the binding processing that happens during monitor creation and updates and resource events. There was also adjust model/DTO and validation changes needed to ensure that users can't improperly combine excludedResourceIds with a resourceId selector.

## How to test

The following primary scenarios are covered by the new tests
  - on monitor update, new exclusions
  - on monitor update, removed exclusions
  - on monitor update, new labels select resource that should be excluded
  - on monitor update, new labels select resource that should be excluded
  - on resource event
  - invalid update, was using resource ID selector, now adding exclusions
  - invalid update, was using exclusions, now adding resource ID selector

Speaking of which, all of the new tests were added to a new test file to avoid further bloating the infamous `MonitorManagementTest` file. I also wanted to make use of the embedded database rather than mock all the repository interactions.

# Depends on

https://github.com/racker/salus-telemetry-model/pull/96
